File: src/lib/dma_buffers_manager.cpp
#include "dma_buffers_manager.h"
#include "v4l2_device.h"
#include <iostream>
#include <optional>
#include <linux/videodev2.h>

DmaBuffersManager::DmaBuffersManager(std::shared_ptr<DmaBufAllocator> allocator, size_t count, v4l2_buf_type type)
    : allocator_(std::move(allocator)), count_(count), type_(type) {
    buffers_.reserve(count_);
    in_use_.resize(count_, false);
}

DmaBuffersManager::~DmaBuffersManager() {
    deallocate();
}

bool DmaBuffersManager::requestOnDevice(V4L2Device& device) {
    struct v4l2_requestbuffers req = {};
    req.count = count_;
    req.type = type_;
    req.memory = V4L2_MEMORY_DMABUF;
    if (!device.request_buffers(req)) {
        std::cerr << "Ошибка запроса " << (type_ == V4L2_BUF_TYPE_VIDEO_OUTPUT_MPLANE ? "входных" : "выходных") 
                  << " DMA-buf буферов" << std::endl;
        return false;
    }
    return true;
}

bool DmaBuffersManager::releaseOnDevice(V4L2Device& device) {
    struct v4l2_requestbuffers req = {};
    req.count = 0;
    req.type = type_;
    req.memory = V4L2_MEMORY_DMABUF;
    // Мы не проверяем результат, так как это делается при очистке
    (void)device.request_buffers(req);
    return true;
}

bool DmaBuffersManager::allocate(size_t buffer_size) {
    if (!allocator_) {
        std::cerr << "DmaBufAllocator не инициализирован" << std::endl;
        return false;
    }

    deallocate(); // Освобождаем старые буферы перед выделением новых
    buffers_.resize(count_);
    in_use_.assign(count_, false);
    current_buffer_ = 0;

    for (size_t i = 0; i < count_; ++i) {
        buffers_[i] = allocator_->allocate(buffer_size);
        if (buffers_[i].fd < 0) {
            std::cerr << "Ошибка выделения DMA-buf буфера " << i << std::endl;
            deallocate(); // Очистка в случае ошибки
            return false;
        }
        if (!allocator_->map(buffers_[i])) {
            std::cerr << "Ошибка маппинга DMA-buf буфера " << i << std::endl;
            deallocate();
            return false;
        }
    }
    return true;
}

void DmaBuffersManager::deallocate() {
    if (!allocator_) return;

    for (auto& dmabuf : buffers_) {
        if (dmabuf.mapped_addr) {
            allocator_->unmap(dmabuf);
            dmabuf.mapped_addr = nullptr;
        }
        if (dmabuf.fd >= 0) {
            allocator_->deallocate(dmabuf);
            dmabuf.fd = -1;
        }
    }
    buffers_.clear();
    in_use_.clear();
}

const DmaBufAllocator::DmaBufInfo& DmaBuffersManager::get_info(size_t index) const {
    return buffers_.at(index);
}

DmaBufAllocator::DmaBufInfo& DmaBuffersManager::get_info(size_t index) {
    return buffers_.at(index);
}

int DmaBuffersManager::get_free_buffer_index() {
    for (size_t i = 0; i < count_; ++i) {
        size_t idx = (current_buffer_ + i) % count_;
        if (!in_use_[idx]) {
            // Не инкрементируем current_buffer_ здесь, чтобы можно было "подсмотреть"
            // свободный буфер, не занимая его.
            return idx;
        }
    }
    return -1; // Нет свободных буферов
}

void DmaBuffersManager::mark_in_use(size_t index) {
    if (index < count_) {
        in_use_[index] = true;
        // Обновляем current_buffer_ только когда буфер действительно используется
        if (index == (current_buffer_ % count_)) {
             current_buffer_ = (index + 1) % count_;
        }
    }
}

void DmaBuffersManager::mark_free(size_t index) {
    if (index < count_) {
        in_use_[index] = false;
    }
}

void DmaBuffersManager::reset_usage() {
    in_use_.assign(count_, false);
    current_buffer_ = 0;
}

File: cs-src/RtpDrmPlayer/Core/DmaBuffersManager.cs
using System;
using System.Collections.Generic;
using System.Runtime.CompilerServices;
using RtpDrmPlayer.Native;

namespace RtpDrmPlayer.Core;

/// <summary>
/// Управляет буферами DMA для V4L2 устройства.
/// Полностью соответствует реализации DmaBuffersManager из C++ версии.
/// </summary>
public class DmaBuffersManager : IDisposable
{
    private readonly DmaBufAllocator _allocator;
    private readonly List<DmaBufAllocator.DmaBufInfo> _buffers = new();
    private readonly bool[] _inUse;
    private readonly int _count;
    private readonly uint _type;
    private int _currentBuffer;
    private bool _disposed;

    public DmaBuffersManager(DmaBufAllocator allocator, int count, uint type)
    {
        Console.WriteLine($"Creating DmaBuffersManager with {count} buffers of type {type}");
        _allocator = allocator;
        _count = count;
        _type = type;
        _inUse = new bool[count];
        _currentBuffer = 0;
    }

    ~DmaBuffersManager()
    {
        Dispose(false);
    }

    public void Dispose()
    {
        Dispose(true);
        GC.SuppressFinalize(this);
    }

    protected virtual void Dispose(bool disposing)
    {
        if (_disposed)
        {
            return;
        }

        if (disposing)
        {
            // Освобождаем управляемые ресурсы
        }

        // Освобождаем неуправляемые ресурсы
        deallocate();

        _disposed = true;
    }

    /// <summary>
    /// Выделяет buffers_.size() буферов указанного размера
    /// </summary>
    /// <param name="bufferSize">Размер каждого буфера в байтах</param>
    /// <returns>true, если все буферы успешно выделены</returns>
    public bool allocate(ulong bufferSize)
    {
        Console.WriteLine($"Allocating {_count} buffers of size {bufferSize} bytes");
        
        // Освобождаем существующие буферы перед выделением новых
        deallocate();
        
        for (int i = 0; i < _count; i++)
        {
            var bufInfo = _allocator.Allocate(bufferSize);
            if (bufInfo.Fd < 0)
            {
                Console.WriteLine($"Failed to allocate buffer {i}");
                deallocate();
                return false;
            }
            
            if (!_allocator.Map(bufInfo))
            {
                Console.WriteLine($"Failed to map buffer {i}");
                deallocate();
                return false;
            }
            
            _buffers.Add(bufInfo);
            Console.WriteLine($"Allocated buffer {i} with fd {bufInfo.Fd}, size {bufInfo.Size}, addr {bufInfo.Mapped}");
        }
        
        Console.WriteLine($"Successfully allocated {_count} buffers");
        return true;
    }

    /// <summary>
    /// Освобождает все буферы и сбрасывает состояние использования
    /// </summary>
    public void deallocate()
    {
        foreach (var buffer in _buffers)
        {
            // Добавляем проверки как в C++ версии для большей надежности
            if (buffer.Mapped != IntPtr.Zero)
            {
                _allocator.Unmap(buffer);
            }
            if (buffer.Fd >= 0)
            {
                _allocator.Deallocate(buffer);
            }
        }
        
        _buffers.Clear();
        Array.Clear(_inUse, 0, _inUse.Length);
        _currentBuffer = 0;
        
        Console.WriteLine("All buffers deallocated");
    }

    /// <summary>
    /// Получает количество буферов
    /// </summary>
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public int count() => _count;
    
    /// <summary>
    /// Получает количество буферов (С# свойство)
    /// </summary>
    public int Count => count();

    /// <summary>
    /// Получает информацию о буфере по индексу
    /// </summary>
    /// <param name="index">Индекс буфера</param>
    /// <returns>Информация о буфере</returns>
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public DmaBufAllocator.DmaBufInfo get_info(int index) => _buffers[index];

    /// <summary>
    /// Индексатор для получения информации о буфере (C# стиль)
    /// </summary>
    public DmaBufAllocator.DmaBufInfo this[int index] => get_info(index);

    /// <summary>
    /// Находит свободный буфер, начиная с текущей позиции
    /// </summary>
    /// <returns>Индекс свободного буфера или -1, если все буферы используются</returns>
    public int get_free_buffer_index()
    {
        for (int i = 0; i < _count; i++)
        {
            int index = (_currentBuffer + i) % _count;
            if (!_inUse[index])
            {
                return index;
            }
        }
        
        Console.WriteLine("No free buffers available");
        return -1;
    }

    /// <summary>
    /// Отмечает буфер как используемый
    /// </summary>
    /// <param name="index">Индекс буфера</param>
    public void mark_in_use(int index)
    {
        if (index < 0 || index >= _count)
        {
            Console.WriteLine($"Invalid buffer index: {index}");
            return;
        }
        
        _inUse[index] = true;
        
        // Если текущий буфер только что был отмечен как используемый,
        // продвигаем указатель на следующий буфер
        if (index == (_currentBuffer % _count))
        {
            _currentBuffer = (index + 1) % _count;
        }
    }

    /// <summary>
    /// Отмечает буфер как свободный
    /// </summary>
    /// <param name="index">Индекс буфера</param>
    public void mark_free(int index)
    {
        if (index < 0 || index >= _count)
        {
            Console.WriteLine($"Invalid buffer index: {index}");
            return;
        }
        
        _inUse[index] = false;
    }

    /// <summary>
    /// Сбрасывает состояние использования всех буферов
    /// </summary>
    public void reset_usage()
    {
        Array.Clear(_inUse, 0, _inUse.Length);
        _currentBuffer = 0;
    }

    /// <summary>
    /// Запрашивает буферы на устройстве V4L2
    /// </summary>
    /// <param name="device">V4L2 устройство</param>
    /// <returns>true, если запрос был успешным</returns>
    public bool requestOnDevice(V4L2Device device)
    {
        Console.WriteLine($"Requesting {_count} buffers on device for type {_type}");
        
        var request = new v4l2_requestbuffers
        {
            count = (uint)_count,
            type = _type,
            memory = V4L2Const.V4L2_MEMORY_DMABUF
        };
        
        bool result = device.RequestBuffers(ref request);
        
        if (!result)
        {
            Console.WriteLine("Failed to request buffers on device");
        }
        else
        {
            Console.WriteLine($"Successfully requested {request.count} buffers on device");
        }
        
        return result;
    }

    /// <summary>
    /// Освобождает буферы на устройстве V4L2
    /// </summary>
    /// <param name="device">V4L2 устройство</param>
    /// <returns>true, если операция была успешной</returns>
    public bool releaseOnDevice(V4L2Device device)
    {
        Console.WriteLine($"Releasing buffers on device for type {_type}");
        
        var request = new v4l2_requestbuffers
        {
            count = 0,
            type = _type,
            memory = V4L2Const.V4L2_MEMORY_DMABUF
        };
        
        bool result = device.RequestBuffers(ref request);
        
        if (!result)
        {
            Console.WriteLine("Failed to release buffers on device");
        }
        else
        {
            Console.WriteLine("Successfully released buffers on device");
        }
        
        return true;  // Всегда возвращаем true как в C++ версии
    }
    
    // Методы совместимости с предыдущей C# версией
    public bool Allocate(ulong size) => allocate(size);
    public void Deallocate() => deallocate();
    public int GetFree() => get_free_buffer_index();
    public void MarkInUse(int index) => mark_in_use(index);
    public void MarkFree(int index) => mark_free(index);
    public bool RequestOnDevice(V4L2Device device, uint bufType) => requestOnDevice(device);
    public bool ReleaseOnDevice(V4L2Device device, uint bufType) => releaseOnDevice(device);

    /// <summary>
    /// Получает тип буфера
    /// </summary>
    public uint Type => _type;
}
